OverloadMethodsDeclarationOrderCheck: split top-level overloads not flagged in compact source files (JEP 512)

Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/coding/overloadmethodsdeclarationorder.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="OverloadMethodsDeclarationOrder"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputOverloadMethodsCompact.java <<'EOF'
void process(int x) { }
void helper() { }
void process(String s) { }
void main() { }
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.5.0-SNAPSHOT-all.jar -c config.xml InputOverloadMethodsCompact.java

Starting audit...
Audit done.
```

Same code inside an ordinary class:

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > OrdinaryOverloadMethods.java <<'EOF'
public class OrdinaryOverloadMethods {
    void process(int x) { }
    void helper() { }
    void process(String s) { }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.5.0-SNAPSHOT-all.jar -c config.xml OrdinaryOverloadMethods.java

Starting audit...
[ERROR] OrdinaryOverloadMethods.java:4:5: All overloaded methods should be placed next to each other. Previous overloaded method located at line '2'. [OverloadMethodsDeclarationOrder]
Audit done.
Checkstyle ends with 1 errors.
```

---

**Describe what you expect in detail.**

OverloadMethodsDeclarationOrderCheck must flag overloaded top-level methods that are not grouped together in a JEP 512 compact source file. The check's `getRequiredTokens` returns only `OBJBLOCK`, and it iterates that node's children looking for `METHOD_DEF` nodes with the same name. A compact source file's implicit class is not represented by an `OBJBLOCK`, so the check never visits the file's top-level methods and overload grouping is never validated.

Expected on the compact input: 1 violation `All overloaded methods should be placed next to each other. Previous overloaded method located at line '1'.` on the second `process(String s)`, identical to the ordinary class above

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheck.java
@@ -82,14 +82,13 @@ public class OverloadMethodsDeclarationOrderCheck extends AbstractCheck {
     public int[] getRequiredTokens() {
         return new int[] {
             TokenTypes.OBJBLOCK,
+            TokenTypes.COMPACT_COMPILATION_UNIT,
         };
     }
 
     @Override
     public void visitToken(DetailAST ast) {
-        final int parentType = ast.getParent().getType();
-
-        final int[] tokenTypes = {
+        final int[] objectBlockParentTypes = {
             TokenTypes.CLASS_DEF,
             TokenTypes.ENUM_DEF,
             TokenTypes.INTERFACE_DEF,
@@ -97,7 +96,8 @@ public class OverloadMethodsDeclarationOrderCheck extends AbstractCheck {
             TokenTypes.RECORD_DEF,
         };
 
-        if (TokenUtil.isOfType(parentType, tokenTypes)) {
+        if (ast.getType() == TokenTypes.COMPACT_COMPILATION_UNIT
+            || TokenUtil.isOfType(ast.getParent().getType(), objectBlockParentTypes)) {
             checkOverloadMethodsGrouping(ast);
         }
     }
@@ -107,7 +107,8 @@ public class OverloadMethodsDeclarationOrderCheck extends AbstractCheck {
      * separated from each other.
      *
      * @param objectBlock
-     *        is a class, interface or enum object block.
+     *        is a class, interface, enum object block, or a compact
+     *        compilation unit for a JEP 512 compact source file.
      */
     private void checkOverloadMethodsGrouping(DetailAST objectBlock) {
         final int allowedDistance = 1;

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheckTest.java
@@ -116,6 +116,37 @@ public class OverloadMethodsDeclarationOrderCheckTest
     }
 
     @Test
+    public void testCompactSourceFileOverloads() throws Exception {
+        final String[] expected = {
+            "13:1: " + getCheckMessage(MSG_KEY, 8),
+        };
+        verifyWithInlineConfigParser(
+            getNonCompilablePath(
+                "InputOverloadMethodsDeclarationOrderCompactSourceFile.java"),
+            expected);
+    }
+
+    @Test
+    public void testCompactSourceFileValid() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+            getNonCompilablePath(
+                "InputOverloadMethodsDeclarationOrderCompactSourceFileValid.java"),
+            expected);
+    }
+
+    @Test
+    public void testCompactSourceFileFirstMethodOverload() throws Exception {
+        final String[] expected = {
+            "11:1: " + getCheckMessage(MSG_KEY, 8),
+        };
+        verifyWithInlineConfigParser(
+            getNonCompilablePath(
+                "InputOverloadMethodsDeclarationOrderCompactSourceFileFirstMethodOverload.java"),
+            expected);
+    }
+
+    @Test
     public void testTokensNotNull() {
         final OverloadMethodsDeclarationOrderCheck check =
             new OverloadMethodsDeclarationOrderCheck();

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/overloadmethodsdeclarationorder/InputOverloadMethodsDeclarationOrderCompactSourceFile.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/overloadmethodsdeclarationorder/InputOverloadMethodsDeclarationOrderCompactSourceFile.java
@@ -1,0 +1,15 @@
+/*
+OverloadMethodsDeclarationOrder
+orderByIncreasingParameterCount = (default)false
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+void process(int x) { }
+
+void helper() { }
+
+// violation below 'All overloaded methods should be placed next to each other.'
+void process(String s) { }
+
+void main() { }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/overloadmethodsdeclarationorder/InputOverloadMethodsDeclarationOrderCompactSourceFileFirstMethodOverload.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/overloadmethodsdeclarationorder/InputOverloadMethodsDeclarationOrderCompactSourceFileFirstMethodOverload.java
@@ -1,0 +1,12 @@
+/*
+OverloadMethodsDeclarationOrder
+orderByIncreasingParameterCount = (default)false
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+void helper() { }
+void process(int x) { }
+void process(String s) { }
+void helper(int x) { } // violation 'All overloaded methods should be placed next to each other.'
+void main() { }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/overloadmethodsdeclarationorder/InputOverloadMethodsDeclarationOrderCompactSourceFileValid.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/overloadmethodsdeclarationorder/InputOverloadMethodsDeclarationOrderCompactSourceFileValid.java
@@ -1,0 +1,11 @@
+/*
+OverloadMethodsDeclarationOrder
+orderByIncreasingParameterCount = (default)false
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+void process(int x) { }
+void process(String s) { }
+void main() { }

